src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/InvalidPiersCheck.java

```diff
@@ -0,0 +1,350 @@
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return an optional {@link CheckFlag} object that
+     */
+    @SuppressWarnings("squid:S3655")
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Edge edge = (Edge) object;
+        // We will mark the edge as flagged as we will only be looking at all way sectioned edges of
+        // a way once
+        this.markAsFlagged(edge.getOsmIdentifier());
+        // Collect all master edges that form the OSM way
+        final Set<Edge> edgesFormingOSMWay = new OsmWayWalker(edge).collectEdges().stream()
+                .filter(Edge::isMasterEdge).collect(Collectors.toSet());
+        final List<Edge> listOfEdgesFormingOSMWay = new ArrayList<>(edgesFormingOSMWay);
+        // We need to sort the edges to build the geometry of the OSM way in a deterministic manner
+        listOfEdgesFormingOSMWay.sort(Comparator.comparingLong(AtlasObject::getIdentifier));
+        // Build a polygon using the locations in the edges
+        final Set<Location> locationsInOsmWay = new HashSet<>();
+        listOfEdgesFormingOSMWay
+                .forEach(osmWayEdge -> locationsInOsmWay.addAll(osmWayEdge.asPolyLine()));
+        final Polygon osmWayAsPolygon = new Polygon(locationsInOsmWay);
+        // Check if the OSM way has linear geometry or polygonal geometry
+        final boolean isPolygonal = this.hasPolygonalGeometry(listOfEdgesFormingOSMWay, edge);
+        final int instructionIndex = isPolygonal ? 1 : 0;
+        // We can flag the edge if it has a highway tag with the right priority or is a polygonal
+        // pier with building tag or is a polygonal pier with amenity=ferry_terminal
+        if ((HighwayTag.highwayTag(edge).isPresent() && HighwayTag.highwayTag(edge).get()
+                .isMoreImportantThanOrEqualTo(this.minimumHighwayTypePier))
+                || (isPolygonal && IS_BUILDING.test(edge))
+                || (isPolygonal && IS_FERRY_TERMINAL.test(edge)))
+        {
+            return Optional.of(this.createFlag(edgesFormingOSMWay,
+                    this.getLocalizedInstruction(instructionIndex, object.getOsmIdentifier())));
+
+        }
+        // Check if the pier has connections to ferry route or buildings
+        final boolean isConnectedToFerryOrBuilding = this.isConnectedToFerryOrBuilding(edge,
+                listOfEdgesFormingOSMWay, isPolygonal, osmWayAsPolygon);
+        // Check if the pier overlaps a highway or not
+        final boolean overlapsHighway = this.pierOverlapsHighway(edge, listOfEdgesFormingOSMWay,
+                osmWayAsPolygon, isPolygonal);
+        // Flag the pier if it overlaps a highway or is connected
+        // to a building or ferry route or overlaps a building
+        return overlapsHighway || isConnectedToFerryOrBuilding
```

Comment:
Do you mind calling out "overlaps highway" and "is connected to very or building" in the instructions ?

Reply:
@danielduhh I had considered adding these in the instructions before but later on decided to drop it because we are flagging piers if
1) they have a highway tag or
2) is connected to ferry/building or
3) overlaps highway or
4) is polygonal pier with building tag or
5) is polygonal pier with amenity=ferry_terminal.
Essentially 5 different cases. There could be cases where two or more of these are true which would make the instructions very long. Also since we are trying to find linear piers and polygonal piers with no area tags, the specifics of which of the above conditions the pier meet, may be irrelevant to editors.

Reply:
I think having 5 use cases is more reason to have specific instructions. With that, people could filter the flags by some string value in the instructions if they only cared about a single case. Thoughts?

Reply:
Sounds reasonable. I will update the instructions likewise.